fineract-provider/src/main/java/org/apache/fineract/portfolio/client/api/ClientsApiResource.java

```diff
@@ -436,4 +352,251 @@ public String retrieveTransferTemplate(@PathParam("clientId") final Long clientI
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "OK", content = @Content(schema = @Schema(implementation = ClientsApiResourceSwagger.PostClientsClientIdResponse.class))) })
+    public String applyCommand(@PathParam("externalId") @Parameter(description = "externalId") final String externalId,
+            @QueryParam("command") @Parameter(description = "command") final String commandParam,
+            @Parameter(hidden = true) final String apiRequestBodyAsJson) {
+        ExternalId clientExternalId = ExternalIdFactory.produce(externalId);
+        final Long clientId = getResolvedClientId(null, clientExternalId);
+
+        final CommandWrapperBuilder builder = new CommandWrapperBuilder().withJson(apiRequestBodyAsJson);
+        final CommandWrapper commandRequest = evaluateCommand(clientId, commandParam, builder);
+        CommandProcessingResult result = commandsSourceWritePlatformService.logCommandSource(commandRequest);
+        return toApiJsonSerializer.serialize(result);
+    }
+
+    @DELETE
+    @Path("/external-id/{externalId}")
+    @Consumes({ MediaType.APPLICATION_JSON })
+    @Produces({ MediaType.APPLICATION_JSON })
+    @Operation(summary = "Delete a Client", description = "If a client is in Pending state, you are allowed to Delete it. The delete is a 'hard delete' and cannot be recovered from. Once clients become active or have loans or savings associated with them, you cannot delete the client but you may Close the client if they have left the program.")
+    @RequestBody(required = true, content = @Content(schema = @Schema(implementation = ClientsApiResourceSwagger.DeleteClientsClientIdRequest.class)))
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "OK", content = @Content(schema = @Schema(implementation = ClientsApiResourceSwagger.DeleteClientsClientIdResponse.class))) })
+    public String delete(@PathParam("externalId") @Parameter(description = "externalId") final String externalId) {
+        ExternalId clientExternalId = ExternalIdFactory.produce(externalId);
+        final Long clientId = getResolvedClientId(null, clientExternalId);
+
+        final CommandWrapper commandRequest = new CommandWrapperBuilder() //
+                .deleteClient(clientId) //
+                .build(); //
+
+        final CommandProcessingResult result = commandsSourceWritePlatformService.logCommandSource(commandRequest);
+
+        return toApiJsonSerializer.serialize(result);
+    }
+
+    @GET
+    @Path("/external-id/{externalId}/obligeedetails")
+    @Consumes({ MediaType.APPLICATION_JSON })
+    @Produces({ MediaType.APPLICATION_JSON })
+    public String retrieveObligeeDetails(@PathParam("externalId") final String externalId, @Context final UriInfo uriInfo) {
+        context.authenticatedUser().validateHasReadPermission(ClientApiConstants.CLIENT_RESOURCE_NAME);
```

Comment:
You could extract the similar business logic into a method and that way we dont need to "nearly" duplicate every API logic here.

Reply:
Done!